python/ray/tune/trial_runner.py

```diff
@@ -944,54 +944,29 @@ def _process_trial_result(self, trial, result):
         flat_result = flatten_dict(result)
         self._validate_result_metrics(flat_result)
 
-        _trigger_callback_complete = False
-
         if self._stopper(trial.trial_id,
                          result) or trial.should_stop(flat_result):
-            result.update(done=True)
-
-            # Hook into scheduler
-            self._scheduler_alg.on_trial_complete(self, trial, flat_result)
-            self._search_alg.on_trial_complete(
-                trial.trial_id, result=flat_result)
-
-            # If this is not a duplicate result, the callbacks should
-            # be informed about the result.
-            if not is_duplicate:
-                with warn_if_slow("callbacks.on_trial_result"):
-                    self._callbacks.on_trial_result(
-                        iteration=self._iteration,
-                        trials=self._trials,
-                        trial=trial,
-                        result=result.copy())
-
-            _trigger_callback_complete = True
             decision = TrialScheduler.STOP
         else:
             with warn_if_slow("scheduler.on_trial_result"):
                 decision = self._scheduler_alg.on_trial_result(
                     self, trial, flat_result)
-            if decision == TrialScheduler.STOP:
-                result.update(done=True)
+        if decision == TrialScheduler.STOP:
+            result.update(done=True)
+        else:
+            # Only updating search alg if the trial is not to be stopped.
```

Comment:
Does this mean we are losing the last result if the stopper decides to stop?

Reply:
I guess the real question is if this impacts searcher performance. Assume we only report once per trial. If this result is not passed to the search algorithm, how can it make better suggestions?

Reply:
So previously we do update the last result _sometimes_, specifically when this line is False:

```
        if self._stopper(trial.trial_id,
                         result) or trial.should_stop(flat_result):
```
and the _scheduler_ decides to stop the trial.

```
        else:
            with warn_if_slow("scheduler.on_trial_result"):
                decision = self._scheduler_alg.on_trial_result(
                    self, trial, flat_result)
            if decision == TrialScheduler.STOP:
                result.update(done=True)
            with warn_if_slow("search_alg.on_trial_result"):
                self._search_alg.on_trial_result(trial.trial_id, flat_result)
```

After reviewing the suggestions, it's probably not that bad. The contract for searchers is that `on_trial_complete` is the important part - intermediate results could be used by some algorithms, but they will always be informed here by the trial complete callback.

So in that case I think it's actually better to remove the on_trial_result call as you did. Let's keep this section in this PR as is